src/Illuminate/Support/LazyCollection.php

```diff
@@ -0,0 +1,1420 @@
+<?php
+
+namespace Illuminate\Support;
+
+use Closure;
+use stdClass;
+use ArrayIterator;
+use IteratorAggregate;
+use Illuminate\Support\Traits\Macroable;
+use Illuminate\Support\Traits\EnumeratesValues;
+
+class LazyCollection implements Enumerable
+{
+    use EnumeratesValues, Macroable;
+
+    /**
+     * The source from which to generate items.
+     *
+     * @var callable|static
+     */
+    public $source;
+
+    /**
+     * Create a new lazy collection instance.
+     *
+     * @param  mixed  $source
+     * @return void
+     */
+    public function __construct($source = null)
+    {
+        if ($source instanceof Closure || $source instanceof self) {
+            $this->source = $source;
+        } elseif (is_null($source)) {
+            $this->source = static::empty();
+        } else {
+            $this->source = static::fromArray($this->getArrayableItems($source));
+        }
+    }
+
+    /**
+     * Create a new instance with no items.
+     *
+     * @return static
+     */
+    public static function empty()
+    {
+        return static::fromArray([]);
+    }
+
+    /**
+     * Create a new instance from the given array.
+     *
+     * @param  array  $array
+     * @return static
+     */
+    public static function fromArray(array $array)
+    {
+        return new static(function () use ($array) {
+            yield from $array;
+        });
+    }
+
+    /**
+     * Create a new instance by invoking the callback a given amount of times.
+     *
+     * @param  int  $number
+     * @param  callable  $callback
+     * @return static
+     */
+    public static function times($number, callable $callback = null)
+    {
+        if ($number < 1) {
+            return new static;
+        }
+
+        $instance = new static(function () use ($number) {
+            for ($current = 1; $current <= $number; $current++) {
+                yield $current;
+            }
+        });
+
+        return is_null($callback) ? $instance : $instance->map($callback);
```

Comment:
```suggestion
        return null === $callback ? $instance : $instance->map($callback);
```

Reply:
Just a suggestion - can we stop with the code reviews suggesting micro optimizations? This PR is already huge with so much effort and work gone into it suggesting a micro optimization such as replacing `is_null($callback)` with `null === $callback` isn't helpful. A seperate PR can be made to do this if anyone would like to.

As I'm sure you can appreciate, contributing to a framework takes a lot of time and effort so we should focus on the actual implementation as a whole rather than nit pick details that really don't make much difference to the overall end goal of this PR.

Reply:
Ok. 

I did not mean to offend you guys or the author. Sorry if this was the case.